File: 백준/Gold/4195. 친구 네트워크/친구 네트워크.py
import sys
T = int(input())
arr = []
friendnum = []
def find_parent(x):
    if arr[x] == x: return x
    arr[x] = find_parent(arr[x])
    return arr[x]

def union(x, y):
    a, b = find_parent(x), find_parent(y)
    if a == b: return
    if a > b:
        friendnum[b] += friendnum[a]
        arr[a] = b
    else:
        friendnum[a] += friendnum[b]
        arr[b] = a

input = sys.stdin.readline
for _ in range(T):
    N = int(input())
    friend = {}
    flag = 0
    arr = [i for i in range(2*N+1)]
    friendnum = [1 for _ in range(2*N+1)]
    for _ in range(N):
        f1, f2 = input().split()
        if f1 not in friend:
            friend[f1] = flag
            flag += 1
        if f2 not in friend:
            friend[f2] = flag
            flag += 1
        union(friend[f1], friend[f2])
        print(friendnum[find_parent(friend[f1])])

File: 백준/Gold/4195. 친구 네트워크/친구 네트워크.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.StringTokenizer;

public class Main {
	static int[] arr;
	static int[] friendnum;
	
	public static int find_parent(int x) {
		if(arr[x] == x) return x;
		else return arr[x] = find_parent(arr[x]); 
	}
	
	public static void union(int x, int y) {
		int a = find_parent(x), b = find_parent(y);
		if(a==b) return;
		if(a>b) {
			friendnum[b]+= friendnum[a];
			arr[a]=b;
		}
		else {
			friendnum[a]+= friendnum[b];
			arr[b]=a; 
		}
	}
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder answer = new StringBuilder();
		int T = Integer.parseInt(br.readLine());
		String f1, f2;
		for(int k=0; k<T; k++) {
			int N = Integer.parseInt(br.readLine());
			arr = new int[2*N+1];
			friendnum = new int[2*N+1];
			for(int i=0; i<2*N+1; i++) {arr[i]= i; friendnum[i] = 1;}  
			int flag = 0;
			HashMap<String, Integer> friend = new HashMap<String, Integer>();
			for(int j=0; j<N; j++) {
				StringTokenizer st = new StringTokenizer(br.readLine());
				f1 = st.nextToken(); f2 = st.nextToken();
				if(!friend.containsKey(f1)) friend.put(f1, flag++);
				if(!friend.containsKey(f2)) friend.put(f2, flag++);
				union(friend.get(f1), friend.get(f2));
				answer.append(friendnum[find_parent(friend.get(f1))]+"\n");
			}
		}
		System.out.print(answer.toString());
	}
}
